core/src/main/scala/org/typelevel/ci/CIString.scala

```diff
@@ -22,54 +22,30 @@ import java.io.Serializable
 import org.typelevel.ci.compat._
 import scala.math.Ordered
 
-/** A case-insensitive String.
-  *
-  * Two CI strings are equal if and only if they are the same length, and each corresponding
-  * character is equal after calling either `toUpper` or `toLower`.
-  *
-  * Ordering is based on a string comparison after folding each character to uppercase and then back
-  * to lowercase.
-  *
-  * All comparisons are insensitive to locales.
-  *
-  * @param toString
-  *   The original value the CI String was constructed with.
-  */
-final class CIString private (override val toString: String)
+@deprecated(
+  message =
+    "Please use either CIStringCF, CIStringCS, or CIStringS instead. CIString/CIStringS implement Unicode default caseless matching on simple case folded strings. For most applications you probably want to use CIStringCF which implements Unicode canonical caseless matching on full case folded strings.",
+  since = "1.3.0")
+final class CIString private (override val toString: String, val asCIStringS: CIStringS)
```

Comment:
Yeah, I'm struggling with what to do here. I'd like `CIString` to be the one most users will want out of the box, but we can't do that unless we break the semantics by changing it to be Unicode canonical caseless matching with full folded strings. 

For most users, they probably won't notice, so maybe it's okay? That also would help with the concern about advertising which is the best type to use by default.

Reply:
Well, are we breaking semantics? Or was our current implementation buggy, and we are fixing the semantics?

The [CI microsite](https://typelevel.org/case-insensitive/) states that "locale independence" is a design goal. Is that what @isomarcte is achieving here?

Reply:
@armanbilge that's a tricky question. Yes and no. The Turkic special cases are obviously not locale independent. However, and forgive me for splitting hairs here, I'm not sure you can say that any Unicode caseless matching is locale independent. Whether or not `"ı" == "I"` in a caseless match is really only answerable in the context of _some_ locale.

In as much as one can have locale independence in this context, the non-Turkic variants achieve that.

I can drop the Turkic ones, but...I don't know...maybe some Turkic language users will want to use them?

Reply:
> Edit: hmm, I see that "locale-independence" is a technical term meaning that e.g. Locale.setDefault(Locale.forLanguageTag("tr")) should not change the result

I did not realize that. :eyes:

Reply:
@armanbilge in that sense then, yes they are all locale independent.

Reply:
Yes, I was a bit muddled until I understood what that meant exactly.

I think the fact that the website demonstrates that:
```scala
assert("ß".toUpperCase == "SS")
assert(CIString("ß") != CIString("SS"))
```
is what locks us into the current semantics, since your proposed changes would affect that right?

Reply:
Yes. That specifically is only valid under simple case folding.

```scala
scala> val a =  "ß"
val a: String = ß

scala> val b = "SS"
val b: String = SS

scala> SimpleCaseFoldedString(a) == SimpleCaseFoldedString(b)
val res7: Boolean = false

scala> FullCaseFoldedString(a) == FullCaseFoldedString(b)                                                                        
val res8: Boolean = true
```

Reply:
Ok. Idk if we can declare this a bug in case-insensitive specifically, but IIUC it is a bug in http4s.

> The [Unicode simple](https://www.w3.org/TR/charmod-norm/#dfn-unicode-simple) casefolding form is not appropriate for string identity matching on the Web.

https://www.w3.org/TR/charmod-norm/#sec_unicode_cs

Reply:
If we did keep around `CIString` and have it target Unicode canonical caseless matching with full folded strings, then I think we'd need to deprecate `length`. The big difference between simple and full folding is that it can change the length of the string, as we see the previous example. I wouldn't want people sizing buffers and such based on `length` when it's not clear _which_ `length` is in play, the original string or the folded one.

Reply:
> Ok. Idk if we can declare this a bug in case-insensitive specifically, but IIUC it is a bug in http4s.
> 
>     The [Unicode simple](https://www.w3.org/TR/charmod-norm/#dfn-unicode-simple) casefolding form is not appropriate for string identity matching on the Web.
> 
> https://www.w3.org/TR/charmod-norm/#sec_unicode_cs

@armanbilge is it clear to you if they are implying default caseless matching or canonical caseless matching? It is not to me.

Reply:
@isomarcte canonical?

> Case sensitivity is not recommended for most specifications but, in the case of an exception where the vocabulary allows non-ASCII characters and which does not want to be sensitive to case distinctions, the ['Unicode canonical case fold'](https://www.w3.org/TR/charmod-norm/#CanonicalFoldNormalizationStep) approach SHOULD be used.

https://www.w3.org/TR/charmod-norm/#matching_CanonicalFoldNormalizationStep

How does "compatibility" fit into this?

> A ['Unicode compatibility case fold'](https://www.w3.org/TR/charmod-norm/#CompatibilityFoldNormalizationStep) approach should not be used.

https://www.w3.org/TR/charmod-norm/#matching_CompatibilityFoldNormalizationStep

Reply:
@armanbilge I don't think compatibility comes into play in our http4s context. (thanks for finding that btw!)

I'm becoming increasingly convinced we should change `CIString` to be full canonical caseless matching and update the docs. I suspect _most_ of our current user base is working around web based technologies and probably should be using full canonical caseless matching anyway.

Reply:
Nothing here contradicts that: https://mvnrepository.com/artifact/org.typelevel/case-insensitive/usages

It's probably fine 🙃

Reply:
I would challenge that this is a bug in http4s.  From [the next paragraph in the spec](https://www.w3.org/TR/charmod-norm/#sec_ascii_cs):

> Specifications that define case-insensitive matching in vocabularies limited to the Basic Latin (ASCII) subset of Unicode MAY specify [ASCII case-insensitive](https://www.w3.org/TR/charmod-norm/#aci) matching.

I can't think of anywhere that http4s is using it that isn't an ASCII context.  The one place I'm less sure of is the URI, but even there, encodings are by convention, so I think the answer is ambiguous.  So, while `CIString` is less general purpose than we would like, is there a smoking gun that its use in the http4s ecosystem is _wrong_?